### ISSUE TITLE
Add memory extension so that we get pointers

### ISSUE THREAD

**Issue by chsasank** [chsasank/llama.lisp#12]
Memory/pointers are described here: https://capra.cs.cornell.edu/bril/lang/memory.html
Tests are here: https://github.com/sampsyo/bril/tree/main/test/interp/mem

Try to get as many tests passed as possible.

**Comment by GlowingScrewdriver** [comment]
Tentatively deciding to use opaque pointers for this. Clang uses opaque pointers, whereas BRIL-LLVM uses transparent pointers and casts.

Documentation on opaque pointers: https://llvm.org/docs/OpaquePointers.html

**Comment by GlowingScrewdriver** [comment]
One difference from Bril: we are going to use the `alloc` instruction for statically allocated arrays on the stack. Bril uses it for dynamically allocated memory on the heap

**Comment by chsasank** [comment]
Alright. Let's only do static allocation. Dynamic stuff can be implemented later with help of a nice runtime library --  including garbage collection eg https://github.com/ivmai/bdwgc.

Our aim is to model LLVM as much as possible with bril/brillisp. Any linker based stuff -- they can come later.

**Comment by chsasank** [comment]
LLVM [supports](https://llvm.org/docs/LangRef.html):

```
Memory Access and Addressing Operations:
    ‘alloca’ Instruction
    ‘load’ Instruction
    ‘store’ Instruction
    ‘fence’ Instruction
    ‘cmpxchg’ Instruction
    ‘atomicrmw’ Instruction
    ‘getelementptr’ Instruction
```

BRIL [supports](https://capra.cs.cornell.edu/bril/lang/memory.html):

```
alloc
free
store
load
ptradd
```

alloc -> alloca
free -> NA
store -> store
load -> load
ptradd -> getelementptr

We'll not implement free for now because llvm doesn't support it. We can add it later as part of runtime.

**Comment by chsasank** [comment]
Let's also not use opaque pointers for now. Let's stick to llvm-15 where typed pointers are still supported.

**Comment by GlowingScrewdriver** [comment]
Started working on this: #15

**Comment by chsasank** [comment]
Closed by #15
